lucille-core/src/main/java/com/kmwllc/lucille/stage/OpenAiEmbedding.java

```diff
@@ -0,0 +1,135 @@
+
+  public OpenAiEmbedding(Config config) throws StageException {
+    super(config, new StageSpec().withRequiredProperties("text_field", "embed_document", "embed_children", "api_key"));
+    textField = config.getString("text_field");
+    embedDocument = config.getBoolean("embed_document");
+    embedChildren = config.getBoolean("embed_children");
+    API_KEY = config.getString("api_key");
+    embeddingField = config.hasPath("embedding_field") ? config.getString("embedding_field") : "embeddings";
+    modelName = OpenAIModels.fromConfig(config);
+    dimensions = config.hasPath("dimensions") ? config.getInt("dimensions") : null;
+    if (!embedDocument && !embedChildren) {
+      throw new StageException("Both embed_document and embed_children are false.");
+    }
+    if (API_KEY.isEmpty()) {
+      throw new StageException("API key is empty.");
+    }
+  }
+
+  @Override
+  public void start() throws StageException {
+    model = OpenAiEmbeddingModel.builder()
+        .modelName(modelName.getModelName())
+        .dimensions(dimensions)
+        .apiKey(API_KEY)
+        .build();
+
+    // retrieve encoding from modelName
+    ModelType modelType = ModelType.fromName(modelName.getModelName()).orElse(ModelType.TEXT_EMBEDDING_3_SMALL);
+    enc = Encodings.newDefaultEncodingRegistry().getEncodingForModel(modelType);
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    if (embedDocument) {
+      sendForEmbedding(doc);
+    }
+    if (embedChildren && doc.hasChildren()){
+      List<Document> childDocs = doc.getChildren();
+      for (Document childDoc : childDocs) {
+        sendForEmbedding(childDoc);
+      }
+    }
+    return null;
+  }
+
+  private void sendForEmbedding(Document doc) throws StageException {
+    if (!doc.has(textField)) {
+      log.warn("{} does not exist, skipping document: {}", textField, doc);
+      return;
+    }
+
+    String content = doc.getString(textField);
+    content = ensureContentIsWithinLimit(content);
+    Response<Embedding> response;
+
+    // model.embed will retry 3 times by default
+    // model.embed might throw runtimeException due to invalid API key and other errors.
+    try {
+      response = model.embed(content);
+    } catch (RuntimeException e) {
+      log.warn("failed to get embedding", e);
+      return;
+    }
+
+    Embedding embedding = response.content();
+    float[] vectors = embedding.vector();
+    for (float vector : vectors) {
+      doc.setOrAdd(embeddingField, vector);
+    }
+  }
+
+  private String ensureContentIsWithinLimit(String content) {
+    int tokenUsage = enc.countTokens(content);
+    String truncatedContent = null;
+    if (tokenUsage > DEFAULT_OPENAI_TOKEN_LIMIT) {
+      EncodingResult encoded = enc.encode(content, DEFAULT_OPENAI_TOKEN_LIMIT);
+      truncatedContent = enc.decode(encoded.getTokens());
+    }
+    return truncatedContent == null ? content : truncatedContent;
```

Comment:
This can be simplified by removing the truncatedContent variable. In the if statement, just `return enc.decode(encoded.getTokens())` and outside of the if just `return content`